File: app/src/main/java/de/blankedv/lanbahnpanel/LocoControlArea.java
package de.blankedv.lanbahnpanel;

import static de.blankedv.lanbahnpanel.LanbahnPanelApplication.*;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;


public final class LocoControlArea {

	private static Paint editPaint, demoPaint;
/*	private static Paint paintLocoAdrTxt, paintLocoSpeedTxt, paintLocoSpeed;
	private static int sliderXoff, sliderYoff;

	private static final float X_LOCO_MID=0.5f;
	private static final float X_LOCO_RANGE=0.25f;

	
	private static final float ZERO_LINE_HALFLENGTH = 15f;  
	private static float canvasWidth = 100f;
	private static long lastSpeedCheckMove=0L;

	private static int sxmin=0, sxmax=0;   */

	private LocoButton stopBtn, commBtn,  clearRoutesBtn; //powerBtn, lampBtn, adrBtn, incrSpeedBtn, decrSpeedBtn, commBtn, powerBtn, functionBtn;

    
    private static float ySpeed = 50f, xSpeed=0;   // needed for position of texts in control area
	
	private static String errorMsg = "";
	private static long errorTime;
	private static long seconds = 0;
	private static long lastSecond = 0;

	//private static float lastXt=X_LOCO_MID;

	//private int selectedLoco=0;

	public LocoControlArea() {
        Paint green, white,

        editPaint  = new Paint();
		editPaint.setColor(Color.RED);
		editPaint.setTextSize(30);
		editPaint.setTypeface(Typeface.DEFAULT_BOLD);
		
		demoPaint  = new Paint();
		demoPaint.setColor(Color.CYAN);
		demoPaint.setTextSize(30);	
		demoPaint.setTypeface(Typeface.DEFAULT_BOLD);

		/* paintLocoSpeedTxt = new Paint();
		paintLocoSpeedTxt.setColor(Color.WHITE);
		paintLocoSpeedTxt.setTextSize(20); */

		green = new Paint();
		green.setColor(Color.GREEN);
		white = new Paint();
		white.setColor(Color.WHITE);

		/*stopBtn = new LocoButton(0.20f,0.5f,
				bitmaps.get("stop_s_on"),
				bitmaps.get("stop_s_off"));
		lampBtn = new LocoButton(0.848f,0.5f,
				bitmaps.get("lamp1"),
				bitmaps.get("lamp0"));
		functionBtn = new LocoButton(0.80f,0.5f,
			 	bitmaps.get("func1"),
			 	bitmaps.get("func0"));  
     	adrBtn = new LocoButton(0.91f,0.5f);  // only text.
		incrSpeedBtn = new LocoButton(0.97f,0.5f,bitmaps.get("incr"));
		decrSpeedBtn = new LocoButton(0.03f,0.5f,bitmaps.get("decr"));  */ 
	
		commBtn = new LocoButton(0.09f,0.5f,bitmaps.get("commok"),bitmaps.get("nocomm"));
		//powerBtn = new LocoButton(0.13f,0.5f, bitmaps.get("greendot"),bitmaps.get("reddot"));
		clearRoutesBtn = new LocoButton(0.25f,0.5f, bitmaps.get("clearrouteson"),bitmaps.get("clearroutesoff"));
	
		
	    //	lonstokeWestBtn = new LocoButton(0.50f,0.6f, bitmaps.get("lonstokewest"));
	}

	public static void dispErrorMsg(String e) {
		errorMsg = e;
		errorTime = System.currentTimeMillis();
	}
	
	public void draw(Canvas canvas) {
			 
		commBtn.doDraw(canvas,LanbahnPanelApplication.connectionIsAlive());
		//powerBtn.doDraw(canvas,LanbahnPanelApplication.isPowerOn());
		if (enableRoutes) {
		   clearRoutesBtn.doDrawBlink(canvas,clearRouteButtonActive); 
		}
		
		for (LampGroup l : lampButtons)
		   l.btn.doDraw(canvas,l.isOn);

		
		canvas.drawBitmap(bitmaps.get("lonstokewest"),(int)(canvas.getWidth()*0.72f), ySpeed*0.22f,  null);
		
		if (enableEdit) canvas.drawText("Edit", (int)(canvas.getWidth()*0.36f), ySpeed*1f, editPaint);
		if (demoFlag) canvas.drawText("Demo", (int)(canvas.getWidth()*0.28f), ySpeed*1f, demoPaint);
		if ( (errorMsg.length() > 0) && (System.currentTimeMillis() - errorTime) < 3000) {
			canvas.drawText(errorMsg, (int)(canvas.getWidth()*0.38f), ySpeed*1f, editPaint);
		} else {
			// clear
			errorMsg = "";
		}
		

		if (System.currentTimeMillis() - lastSecond  >= 1000 )  {
			// increment seconds timer
			seconds++;
			lastSecond = System.currentTimeMillis();
			Route.auto();
		}
		
	}
		// draw "buttons" and states
		/* no loco
		adrBtn.doDraw(canvas,getLocoAdr(), paintLocoAdrTxt);
		lampBtn.doDraw(canvas,((LanbahnPanelApplication.getLocoSXData() & 0x40) != 0));
	 	functionBtn.doDraw(canvas,((LanbahnPanelApplication.getLocoSXData() & 0x80) != 0));
		stopBtn.doDraw(canvas,((LanbahnPanelApplication.getLocoSXData() & 0x1f) != 0));
		incrSpeedBtn.doDraw(canvas);
		decrSpeedBtn.doDraw(canvas);  
		*/
		
		
        /* no loco control	
		 draw slider for speed selection
		sxmin = (int)((canvas.getWidth()*(X_LOCO_MID-X_LOCO_RANGE)));
		sxmax = (int)((canvas.getWidth()*(X_LOCO_MID+X_LOCO_RANGE)));

		Rect speedLine = new Rect(sxmin,(int)(ySpeed-1),sxmax,(int)ySpeed+1);       
		canvas.drawRect(speedLine, white);

		Rect zeroLine = new Rect((sxmin+sxmax)/2-1,(int)(ySpeed-ZERO_LINE_HALFLENGTH),
				(sxmin+sxmax)/2+1,(int)(ySpeed+ZERO_LINE_HALFLENGTH));   
		canvas.drawRect(zeroLine, white);

		canvasWidth = canvas.getWidth();	

		int xtext =  (int)(canvasWidth*(X_LOCO_MID+X_LOCO_RANGE*0.9f));
		canvas.drawText(locoSpeed(),xtext, ySpeed+20, paintLocoSpeedTxt);


		xSpeed = canvasWidth*sxpeed();
		canvas.drawBitmap(bitmaps.get("slider"), xSpeed-sliderXoff, ySpeed-sliderYoff,  null);
		if ( (System.currentTimeMillis() - lastSpeedCheckMove) < 1000) {
			// draw additional "soll" slider
			canvas.drawBitmap(bitmaps.get("slider_grey"), xSpeed-sliderXoff, ySpeed-sliderYoff,  null);
			canvas.drawBitmap(bitmaps.get("slider"), lastXt-sliderXoff, ySpeed-sliderYoff,  null);
		} else {
			canvas.drawBitmap(bitmaps.get("slider"), xSpeed-sliderXoff, ySpeed-sliderYoff,  null);
		}
		
	

	}

	/* disabled loco control
	private float sxpeed() {
		int s = (LanbahnPanelApplication.getLocoSXData() & 0x1f);
		if ((LanbahnPanelApplication.getLocoSXData() & 0x20) != 0) s = -s;
		float speed = (X_LOCO_RANGE*(float)s / 31.0f)+X_LOCO_MID;  // sx=31 ==> ~0.25f (+0.6f offset)
		return speed;
	}

	private  String locoSpeed() {		
		int s = (LanbahnPanelApplication.getLocoSXData() & 0x1f);
		if ((LanbahnPanelApplication.getLocoSXData() & 0x20) != 0) s = -s;
		return ""+s;
	}

	private Boolean forward() {		
		return ( (LanbahnPanelApplication.getLocoSXData() & 0x20) != 0);

	}

	private int speedToSX(int s) {
		int data=0;
		if (s < 0) { data |= 0x20; s= -s;}
		if (s > 31) s=31;
		data |= s;
		return data;
	}

	public void checkSpeedMove(float xt, float yt) {
		// check slider
		//if (DEBUG) Log.d(TAG,"check slider touch xt="+xt+"  yt="+yt+" xSpeed="+xSpeed+" sliderXoff="+sliderXoff+" ySpeed="+ySpeed+" sliderYoff="+sliderYoff);

		if (   (xt > (X_LOCO_MID-X_LOCO_RANGE)*canvasWidth) 
				&& (xt < (X_LOCO_MID+X_LOCO_RANGE)*canvasWidth) 
				&& (yt > (ySpeed-sliderYoff)) 
				&& (yt < (ySpeed+sliderYoff))) {
			lastSpeedCheckMove = System.currentTimeMillis();
			lastXt = xt;
			int s = (int)(((31.4f/X_LOCO_RANGE)*(xt - X_LOCO_MID*canvasWidth))/canvasWidth);
			int sxData = speedToSX(s);
			LanbahnPanelApplication.setLocoSpeed(sxData);  // will be sent only when different turnout currently known speed.
			//LanbahnPanelApplication.sendSpeed(0,true);
		}
	}
	*/
	public void checkTouch(float x, float y) {
		if ( (enableRoutes) && (clearRoutesBtn.isTouched(x, y))) {
			if (!clearRouteButtonActive) {
				clearRouteButtonActive = true;
			} else {
				// actual clear route is done in "RouteButtonElement" class
				clearRouteButtonActive = false;
			}
		} 
		else {
			for (LampGroup l : lampButtons) {
				if (l.btn.isTouched(x,y))	l.toggle();
			}
		}
			
		/*else if (decrSpeedBtn.isTouched(x,y)) {
			LanbahnPanelApplication.decrLocoSpeed();
		} else if (functionBtn.isTouched(x,y)) {
			LanbahnPanelApplication.toggleFunc();
		} else if (adrBtn.isTouched(x,y)) {
			Dialogs.selectLocoAddressDialog();
			
		} */

	} 

	public void recalcGeometry() {
		//stopBtn.recalcXY();
		//lampBtn.recalcXY();
		//adrBtn.recalcXY();	
		//decrSpeedBtn.recalcXY();
		//incrSpeedBtn.recalcXY();
		//functionBtn.recalcXY();
		//powerBtn.recalcXY();
		commBtn.recalcXY();
		clearRoutesBtn.recalcXY();
		
		for (LampGroup l : lampButtons) l.btn.recalcXY();

		//ySpeed=(controlAreaRect.bottom - controlAreaRect.top)/2;  // mid of control area range.
	}

	
}

File: app/src/main/java/de/blankedv/lanbahnpanel/loco/LocoControlArea.kt
package de.blankedv.lanbahnpanel.loco

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color.*
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Typeface
import android.util.Log
import de.blankedv.lanbahnpanel.model.*
import de.blankedv.lanbahnpanel.util.LanbahnBitmaps.bitmaps
import de.blankedv.lanbahnpanel.view.Dialogs

/**
 * handles the display of the top 20% of the display, the "LOCO CONTROL AREA"
 * one loco can be controlled (out of the ones defined in "locos-xyz.xml" File)
 *
 * @author mblank
 */
class LocoControlArea(internal var ctx: Context) {

    private val stopBtn: LocoButton
    private val lampBtn: LocoButton    // F0
    private val addressBtn: LocoButton
    private val functionBtn: LocoButton  //F1
    private val leftBtn: LocoButton  // speed to left
    private val rightBtn: LocoButton  // speed to right

    private var paintText = Paint()
    private var paintLargeTxt = Paint()
    private var paintTextDisabled = Paint()

    private var green = Paint()
    private var white = Paint()
    private var editPaint = Paint()
    private var sliderXoff = bitmaps.get("slider")?.getWidth()!! / 2
    private var sliderYoff = bitmaps.get("slider")?.getHeight()!! / 2

    init {

        val metrics = ctx.resources.displayMetrics
        val width = metrics.widthPixels
        val height = metrics.heightPixels

        // define some paints for the loco controls and texts

        paintLargeTxt.color = WHITE
        paintLargeTxt.textSize = calcTextSize(width)

        paintText.color = WHITE
        paintText.textSize = paintLargeTxt.textSize * 0.7f

        paintTextDisabled.color = LTGRAY
        paintTextDisabled.textSize = paintLargeTxt.textSize * 0.7f


        editPaint.color = RED
        editPaint.textSize = calcTextSize(width)
        editPaint.typeface = Typeface.DEFAULT_BOLD
        green.color = GREEN
        white.color = WHITE

        addressBtn = LocoButton(0.11f, 0.5f)  // only text.
        stopBtn = LocoButton(0.20f, 0.5f)  // only text
        lampBtn = LocoButton(0.848f, 0.5f,
                bitmaps?.get("lamp1")!!,     // bitmap for "on" state
                bitmaps?.get("lamp0")!!)     // bitmap for "off" state
        functionBtn = LocoButton(0.796f, 0.5f,
                bitmaps?.get("func1")!!,
                bitmaps?.get("func0")!!)

        leftBtn = LocoButton(0.03f, 0.5f, bitmaps?.get("left")!!, bitmaps?.get("left")!!)
        rightBtn = LocoButton(0.97f, 0.5f, bitmaps?.get("right")!!)

      }

    private fun drawSlider(canvas : Canvas) {

    }
    fun draw(canvas: Canvas) {

        // Done in activity regularly
        //        selectedLoco?.updateLocoFromSX()  // to be able to display actual states of this loco

        // draw "buttons" and states
        leftBtn.doDraw(canvas,true)
        rightBtn.doDraw(canvas,true)
        addressBtn.doDraw(canvas, "A:"+selectedLoco?.adr!!.toString(), paintLargeTxt)
        lampBtn.doDraw(canvas, selectedLoco?.lamp_to_be!!)
        functionBtn.doDraw(canvas, selectedLoco?.function_to_be!!)
        if (selectedLoco?.speed_act !== 0) {
            stopBtn.doDraw(canvas, "Stop", paintText)
        } else {
            stopBtn.doDraw(canvas, "(Stop)", paintTextDisabled)
        }
        // draw slider for speed selection
        sxmin = (canvas.width * (X_LOCO_MID - X_LOCO_RANGE)).toInt()
        sxmax = (canvas.width * (X_LOCO_MID + X_LOCO_RANGE)).toInt()

        val speedLine = Rect(sxmin, (ySpeed - 1).toInt(), sxmax, ySpeed.toInt() + 1)
        canvas.drawRect(speedLine, white)

        val zeroLine = Rect((sxmin + sxmax) / 2 - 1, (ySpeed - ZERO_LINE_HALFLENGTH).toInt(),
                (sxmin + sxmax) / 2 + 1, (ySpeed + ZERO_LINE_HALFLENGTH).toInt())
        canvas.drawRect(zeroLine, white)

        canvasWidth = canvas.width.toFloat()

        xSpeedAct = canvasWidth * sxSpeed()   // grey slider on bottom
        xSpeedToBe = canvasWidth * speedToBe()  // orange slider on top

        canvas.drawBitmap(bitmaps.get("slider_grey"), xSpeedAct - sliderXoff, ySpeed - sliderYoff, null)
        canvas.drawBitmap(bitmaps.get("slider"), xSpeedToBe - sliderXoff, ySpeed - sliderYoff, null)

        var xtext = (canvasWidth * (X_LOCO_MID + X_LOCO_RANGE * 0.9f)).toInt()
        canvas.drawText(locoSpeed(), xtext.toFloat(), ySpeed + 32, paintText)
        xtext = (canvasWidth * (X_LOCO_MID - X_LOCO_RANGE * 0.9f)).toInt()
        canvas.drawText(selectedLoco?.longString(), xtext.toFloat(), ySpeed + 32, paintText)



    }

    private fun sxSpeed(): Float {
        val s = selectedLoco?.speed_from_sx

        return X_LOCO_RANGE * s?.toFloat()!! / 31.0f + X_LOCO_MID
    }

    private fun speedToBe(): Float {
        val s = selectedLoco?.speed_to_be
        return X_LOCO_RANGE * s?.toFloat()!! / 31.0f + X_LOCO_MID
    }

    private fun locoSpeed(): String {
        val s = selectedLoco?.speed_from_sx
        return "" + s
    }


    fun checkSpeedMove(xt: Float, yt: Float) {
        // check slider
        //if (DEBUG) Log.d(TAG,"check slider touch xt="+xt+"  yt="+yt+" xSpeed="+xSpeed+" sliderXoff="+sliderXoff+" ySpeed="+ySpeed+" sliderYoff="+sliderYoff);

        if (xt > (X_LOCO_MID - X_LOCO_RANGE) * canvasWidth
                && xt < (X_LOCO_MID + X_LOCO_RANGE) * canvasWidth
                && yt > ySpeed - sliderYoff
                && yt < ySpeed + sliderYoff) {
            lastSpeedCheckMove = System.currentTimeMillis()
            lastXt = xt
            val s = Math.round(31.0f / X_LOCO_RANGE * (xt - X_LOCO_MID * canvasWidth) / canvasWidth)
            if (DEBUG) Log.d(TAG, "slider, speed set to be = $s")
            selectedLoco?.setSpeed(s)  // will be sent only when different to currently known speed.

        }
    }

    /**
     * check, if the control area was touched at the Button positions or at the speed slider
     *
     * @param x
     * @param y
     */
    fun checkTouch(x: Float, y: Float) {
        if (stopBtn.isTouched(x, y)) {
            selectedLoco?.stopLoco()
        } else if (lampBtn.isTouched(x, y)) {
            selectedLoco?.toggleLocoLamp()
        } else if (functionBtn.isTouched(x, y)) {
            selectedLoco?.toggleFunc()
        } else if (addressBtn.isTouched(x, y)) {
            Dialogs.selectLocoDialog()
        } else if (leftBtn.isTouched(x, y)) {
            selectedLoco?.decrLocoSpeed()
        } else if (rightBtn.isTouched(x, y)) {
            selectedLoco?.incrLocoSpeed()
        }

    }
/*
    fun checkIncrDecrSpeed(x: Float, y: Float) {
        if (leftBtn.isTouched(x, y)) {
            selectedLoco?.startDecrLocoSpeed()
        } else if (rightBtn.isTouched(x, y)) {
            selectedLoco?.startIncrLocoSpeed()
        }
    } */

    fun recalcGeometry() {
        stopBtn.recalcXY()
        lampBtn.recalcXY()
        addressBtn.recalcXY()
        functionBtn.recalcXY()
        leftBtn.recalcXY()
        rightBtn.recalcXY()
        ySpeed = (controlAreaRect?.bottom!!.toFloat() - controlAreaRect?.top!!) / 2  // mid of control area range.
    }

    private fun calcTextSize(w: Int): Float {
        // text size = 30 for width=1024
        return 30.0f * w / 1024
    }

    companion object {



        private val X_LOCO_MID = 0.5f
        private val X_LOCO_RANGE = 0.25f

        private var ySpeed = 50f
        private var xSpeedAct = 0f
        private var xSpeedToBe = 0f
        private val ZERO_LINE_HALFLENGTH = 15f  //
        private var canvasWidth = 100f

        private var sxmin = 0
        private var sxmax = 0
        private var lastSpeedCheckMove = 0L
        private var lastXt = X_LOCO_MID
    }

}
